develop/tutorials/articles/06-web-applications/05-jsf/01-packaging-a-jsf-application.markdown

```diff
@@ -0,0 +1,161 @@
+  to affect the content of an OSGi-ready `META-INF/MANIFEST.MF`.
+- Can take advantage of OSGi Declarative Services (DS)
+- Provides a class that extends.
+  [GenericPortlet](http://portals.apache.org/pluto/portlet-2.0-apidocs/javax/portlet/GenericPortlet.html)
+  (or in the case of JSF,
+  [GenericFacesPortlet](http://myfaces.apache.org/portlet-bridge/2.0/api/apidocs/javax/portlet/faces/GenericFacesPortlet.html))
+  and specify portlet attributes via the `@Component` annotation, rather than
+  using two separate descriptors: `WEB-INF/portlet.xml` and
+  `WEB-INF/liferay-portlet.xml`.
+
+**Drawbacks:**
+
+- The `bnd.bnd` file must include embedded dependency information that is
+  already present in the `pom.xml`.
+- Bypasses the Liferay auto-deploy process, which means developers must have
+  the `WEB-INF/web.xml` descriptor fully ready for deployment, such as
+  `<servlet>` and `<servlet-mapping>` for Liferay's `PortletServlet` and a
+  `<listener>` for Liferay's `PluginContextListener`.
+
+## Excluding the Bundle-SymbolicName OSGi Directive in Your WAB
+
+**Benefits:**
+
+- Processed by the Liferay auto-deploy process, which adds the aforementioned
+  `PortletServlet` and `PluginContextListener`.
+- Processed by the Liferay
+  [WAB Generator](https://github.com/liferay/liferay-portal/tree/master/modules/apps/foundation/portal-osgi-web/portal-osgi-web-wab-generator),
+  which automatically creates an OSGi-ready `META-INF/MANIFEST.MF`.
+- Can affect the content of `META-INF/MANIFEST.MF` by putting BND directives and
+  OSGi headers into the `WEB-INF/liferay-plugin/packages.properties` file.
+
+**Drawbacks:**
+
+- Can't supply the `bnd.bnd` and can't utilize a build-time plugin such as the
+  [bnd-maven-plugin](http://njbartlett.name/2015/03/27/announcing-bnd-maven-plugin.html).
+- Can't take advantage of OSGi Declarative Services (DS)
+- Developer must specify the `WEB-INF/portlet.xml` and
+  `WEB-INF/liferay-portlet.xml` descriptors.
+
+Both approaches can use custom-built OSGi services, since both result in the
+deployment of a WAB (true OSGi modules). Also, both approaches can be used to
+lookup OSGi services, although including the `Bundle-SymbolicName` OSGi
+directive allows you to take advantage of OSGi Declarative Services via the
+`@Reference` annotation. A JSF-friendly lookup of an OSGi service for both
+approaches can be done by using the JSF Expression Language (EL):
+
+    @ManagedProperty(value = "#{userLocalService}")
+    private UserLocalService userLocalService;
+
+    public void setUserLocalService(UserLocalService userLocalService) {
+        this.userLocalService = userLocalService;
+    }
+
+Then you can write an
+[ELResolver](http://docs.oracle.com/javaee/7/api/javax/el/ELResolver.html) that
+does the following:
+
+    BundleContext bundleContext = FrameworkUtil.getBundle(getClass()).getBundleContext();
+    ServiceReference<?>[] refs = ctx.getServiceReferences(UserLocalService.class.getName(), null);
```

Comment:
Yes. Better to specify UserService because it encourages developers to write code that takes advantage of permissions. UserLocalService bypasses permissions.